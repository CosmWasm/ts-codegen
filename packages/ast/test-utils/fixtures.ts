// @ts-nocheck
import { JSONSchema } from '../src/types';

import canExecuteRelayResponseData from '../../../__fixtures__/vectis/govec/can_execute_relay_response.json';
import cosmosMsgForEmptyData from '../../../__fixtures__/vectis/govec/cosmos_msg_for__empty.json';
import executeMsgForEmptyData from '../../../__fixtures__/vectis/govec/execute_msg_for__empty.json';
import infoResponseData from '../../../__fixtures__/vectis/govec/info_response.json';
import relayTransactionData from '../../../__fixtures__/vectis/govec/relay_transaction.json';

import nftIdlVersionData from '../../../__fixtures__/idl-version/accounts-nft/account-nft.json';

export const nftIdlVersion: JSONSchema = nftIdlVersionData;
export const vectisCanExecuteRelayResponse: JSONSchema = canExecuteRelayResponseData;
export const vectisCosmosMsgForEmpty: JSONSchema = cosmosMsgForEmptyData;
export const vectisExecuteMsgForEmpty: JSONSchema = executeMsgForEmptyData;
export const vectisInfoResponse: JSONSchema = infoResponseData;
export const vectisRelayTransaction: JSONSchema = relayTransactionData;
