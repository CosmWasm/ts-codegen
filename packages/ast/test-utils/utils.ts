import generate from '@babel/generator';
import { RenderOptions } from '@cosmwasm/ts-codegen-ast/types';
import { ExecuteMsg, IDLObject, JSONSchema, QueryMsg } from '@cosmwasm/ts-codegen-types';
import { readFileSync } from 'fs';
import { sync as glob } from 'glob';
import { join } from 'path';

import { RenderContext } from '../src/context';

export const expectCode = (ast: any): void => {
  expect(
    generate(ast).code
  ).toMatchSnapshot();
};

export const printCode = (ast: any): void => {
  console.log(
    generate(ast).code
  );
};

export const makeContext = (
  schema: JSONSchema,
  options?: RenderOptions,
  responses?: Record<string, JSONSchema>
) => {
  return new RenderContext({
    schemas: [schema],
    responses
  }, options);
};

interface GlobContractLegacy {
  name: `/${string}.json`;
  content: JSONSchema;
}

interface GlobContract {
  name: `/${string}.json`;
  content: IDLObject;
}

const fixtureDir = join(__dirname, '../../../__fixtures__');
const globCache: Record<string, GlobContractLegacy[] | GlobContract[]> = {};

export const globIdlBasedContracts = (p: string): GlobContract[] => {
  if (globCache[p]) return globCache[p] as GlobContract[];
  // @ts-ignore
  const contracts: GlobContract[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const globLegacyContracts = (p: string): GlobContractLegacy[] => {
  if (globCache[p]) return globCache[p] as GlobContractLegacy[];
  // @ts-ignore
  const contracts: GlobContractLegacy[] = glob(join(fixtureDir, p, '/*.json')).map(file => {
    return {
      name: file.split(join('__fixtures__', p))[1],
      content: JSON.parse(readFileSync(file, 'utf-8'))
    };
  });
  globCache[p] = contracts;
  return contracts;
};

export const getMsgExecuteLegacyFixture = (
  p: string,
  name: `/${string}.json`
): ExecuteMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as ExecuteMsg;
};

export const getMsgQueryLegacyFixture = (
  p: string,
  name: `/${string}.json`
): QueryMsg => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as QueryMsg;
};

export const getLegacyFixture = (
  p: string,
  name: `/${string}.json`
): JSONSchema => {
  const contracts = globLegacyContracts(p);
  const contract = contracts.find(a => a.name === name);
  return contract.content as JSONSchema;
};
