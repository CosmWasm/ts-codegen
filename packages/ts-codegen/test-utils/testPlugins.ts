import {
  BuilderContext,
  ContractInfo,
  RenderContext,
  RenderContextBase,
  RenderOptions,
} from "wasm-ast-types";
import { ClientPlugin, MessageComposerPlugin, TSBuilderOptions } from "../src";
import { ContractsContextProviderPlugin } from "../src/plugins/provider";
import { ContractsProviderBundlePlugin } from "../src/plugins/provider-bundle";

export class TestClientPlugin extends ClientPlugin {
  builderContext: BuilderContext;

  constructor(opts: RenderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: RenderOptions
  ): RenderContextBase<RenderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestMessageComposerPlugin extends MessageComposerPlugin {
  builderContext: BuilderContext;

  constructor(opts: RenderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: RenderOptions
  ): RenderContextBase<RenderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestContractsContextProviderPlugin extends ContractsContextProviderPlugin {
  builderContext: BuilderContext;

  constructor(opts: TSBuilderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: TSBuilderOptions
  ): RenderContextBase<TSBuilderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}

export class TestContractsProviderBundlePlugin extends ContractsProviderBundlePlugin {
  builderContext: BuilderContext;

  constructor(opts: TSBuilderOptions, builderContext: BuilderContext) {
    super(opts);
    this.builderContext = builderContext;
  }

  initContext(
    contract: ContractInfo,
    options?: TSBuilderOptions
  ): RenderContextBase<TSBuilderOptions> {
    return new RenderContext(contract, options, this.builderContext);
  }
}
