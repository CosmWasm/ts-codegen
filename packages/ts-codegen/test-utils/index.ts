import {
  BuilderContext,
  ContractInfo,
  ReactQueryOptions,
  RenderOptions,
  TSTypesOptions,
} from "wasm-ast-types";
import { ReactQueryPlugin } from "../src/plugins/react-query";
import { TypesPlugin } from "../src/plugins/types";
import { ClientPlugin } from "../src/plugins/client";
import { MessageComposerPlugin } from "../src/plugins/message-composer";
import { RecoilPlugin } from "../src/plugins/recoil";
import { MessageBuilderPlugin } from "../src/plugins/message-builder";
import { ContractsContextProviderPlugin } from "../src/plugins/provider";
import { ContractsProviderBundlePlugin } from "../src/plugins/provider-bundle";
import { TSBuilderOptions } from "../src/builder";
import { createDefaultContractInfo } from "../src/utils/contracts";
import {
  TestContractsContextProviderPlugin,
  TestClientPlugin,
  TestMessageComposerPlugin,
  TestContractsProviderBundlePlugin,
} from "./testPlugins";
import deepmerge from "deepmerge";

export const testDefaultOptions: RenderOptions = {
  enabled: true,
  types: {
    enabled: true,
    itemsUseTuples: false,
    aliasExecuteMsg: false,
  },
  client: {
    enabled: true,
    execExtendsQuery: true,
    noImplicitOverride: false,
  },
  recoil: {
    enabled: true,
  },
  messageComposer: {
    enabled: true,
  },
  messageBuilder: {
    enabled: true,
  },
  reactQuery: {
    enabled: true,
    optionalClient: false,
    version: "v4",
    mutations: false,
    camelize: true,
    queryKeys: false,
  },
  useContractsHook: {
    enabled: true,
  },
};

export async function generateReactQuery(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: ReactQueryOptions
) {
  const options = opts
    ? deepmerge(testDefaultOptions, { reactQuery: opts })
    : testDefaultOptions;

  await new ReactQueryPlugin(options).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateTypes(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: TSTypesOptions
) {
  await new TypesPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateClient(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestClientPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new ClientPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageComposer(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext,
  opts?: RenderOptions
) {
  if (builderContext) {
    await new TestMessageComposerPlugin(
      opts ?? testDefaultOptions,
      builderContext
    ).render(outPath, contractName, contractInfo);
  } else {
    await new MessageComposerPlugin(opts ?? testDefaultOptions).render(
      outPath,
      contractName,
      contractInfo
    );
  }
}

export async function generateMessageBuilder(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new MessageBuilderPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateRecoil(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  opts?: RenderOptions
) {
  await new RecoilPlugin(opts ?? testDefaultOptions).render(
    outPath,
    contractName,
    contractInfo
  );
}

export async function generateContractHooks(
  contractName: string,
  contractInfo: ContractInfo,
  outPath: string,
  builderContext?: BuilderContext
) {
  await new TestContractsContextProviderPlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, contractName, contractInfo);

  await new TestContractsProviderBundlePlugin(
    {
      enabled: true,
      useShorthandCtor: true,
      useContractsHook: {
        enabled: true,
      },
    },
    builderContext
  ).render(outPath, undefined, contractInfo ?? createDefaultContractInfo());
}
