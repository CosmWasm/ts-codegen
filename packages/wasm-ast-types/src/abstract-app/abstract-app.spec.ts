import autocompounder_schema from '../../../../__fixtures__/abstract/apps/autocompounder.json';
import { expectCode, makeContext } from '../../test-utils';
import {
  createAppExecuteClass,
  createAppExecuteInterface,
  createAppQueryClass,
  createAppQueryInterface
} from './abstract-app';

it('IAutocompounderAppQueryClient', () => {
  const ctx = makeContext(autocompounder_schema);

  expectCode(
    createAppQueryInterface(
      ctx,
      'IAutocompounderAppQueryClient',
      'AutocompounderAppClient',
      autocompounder_schema.query
    )
  );
});

it('AutocompounderAppQueryClient', () => {
  const ctx = makeContext(autocompounder_schema);

  expectCode(
    createAppQueryClass(
      ctx,
      'Autocompounder',
      'AutocompounderAppQueryClient',
      'IAutocompounderAppQueryClient',
      autocompounder_schema.query
    )
  );
});

it('IAutocompounderAppClient', () => {
  const ctx = makeContext(autocompounder_schema);

  expectCode(
    createAppExecuteInterface(
      ctx,
      'IAutocompounderAppClient',
      'AutocompounderAppClient',
      'IAutocompounderAppQueryClient',
      autocompounder_schema.execute
    )
  );
});

it('AutocompounderAppClient', () => {
  const ctx = makeContext(autocompounder_schema);

  expectCode(
    createAppExecuteClass(
      ctx,
      'Autocompounder',
      'AutocompounderAppClient',
      'IAutocompounderAppClient',
      'AutocompounderAppQueryClient',
      autocompounder_schema.execute
    )
  );
});
