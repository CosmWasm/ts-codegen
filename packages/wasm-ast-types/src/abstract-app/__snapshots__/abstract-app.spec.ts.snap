// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompounderAppClient 1`] = `
"export class AutocompounderAppClient extends AutocompounderAppQueryClient implements IAutocompounderAppClient {
  accountClient: AbstractAccountClient;

  constructor({
    abstractClient,
    accountId,
    managerAddress,
    proxyAddress,
    moduleId
  }: {
    abstractClient: AbstractClient;
    accountId: number;
    managerAddress: string;
    proxyAddress: string;
    moduleId: string;
  }) {
    super({
      abstractQueryClient: abstractClient,
      accountId,
      managerAddress,
      proxyAddress,
      moduleId
    });
    this.accountClient = AbstractAccountClient.fromQueryClient(this.accountQueryClient, abstractClient);
    this.updateFeeConfig = this.updateFeeConfig.bind(this);
    this.deposit = this.deposit.bind(this);
    this.withdraw = this.withdraw.bind(this);
    this.compound = this.compound.bind(this);
    this.batchUnbond = this.batchUnbond.bind(this);
  }

  updateFeeConfig = async (params: CamelCasedProperties<Extract<ExecuteMsg, {
    update_fee_config: unknown;
  }>[\\"update_fee_config\\"]>, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return this._execute(AutocompounderExecuteMsgBuilder.updateFeeConfig(params), fee, memo, _funds);
  };
  deposit = async (params: CamelCasedProperties<Extract<ExecuteMsg, {
    deposit: unknown;
  }>[\\"deposit\\"]>, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return this._execute(AutocompounderExecuteMsgBuilder.deposit(params), fee, memo, _funds);
  };
  withdraw = async (fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return this._execute(AutocompounderExecuteMsgBuilder.withdraw(), fee, memo, _funds);
  };
  compound = async (fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return this._execute(AutocompounderExecuteMsgBuilder.compound(), fee, memo, _funds);
  };
  batchUnbond = async (fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    return this._execute(AutocompounderExecuteMsgBuilder.batchUnbond(), fee, memo, _funds);
  };
  _execute = async (msg: ExecuteMsg, fee: number | StdFee | \\"auto\\" = \\"auto\\", memo?: string, _funds?: Coin[]): Promise<ExecuteResult> => {
    const moduleMsg: AppExecuteMsg<ExecuteMsg> = AppExecuteMsgFactory.executeApp(msg);
    return await this.accountClient.abstract.client.execute(this.accountClient.sender, await this.getAddress(), moduleMsg, fee, memo, _funds);
  };
}"
`;

exports[`AutocompounderAppQueryClient 1`] = `
"export class AutocompounderAppQueryClient implements IAutocompounderAppQueryClient {
  accountQueryClient: AbstractAccountQueryClient;
  moduleId: string;
  _moduleAddress: string | undefined;

  constructor({
    abstractQueryClient,
    accountId,
    managerAddress,
    proxyAddress,
    moduleId
  }: {
    abstractQueryClient: AbstractQueryClient;
    accountId: number;
    managerAddress: string;
    proxyAddress: string;
    moduleId: string;
  }) {
    this.accountQueryClient = new AbstractAccountQueryClient({
      abstract: abstractQueryClient,
      accountId,
      managerAddress,
      proxyAddress
    });
    this.moduleId = moduleId;
    this.config = this.config.bind(this);
    this.pendingClaims = this.pendingClaims.bind(this);
    this.claims = this.claims.bind(this);
    this.allClaims = this.allClaims.bind(this);
    this.latestUnbonding = this.latestUnbonding.bind(this);
    this.totalLpPosition = this.totalLpPosition.bind(this);
    this.balance = this.balance.bind(this);
  }

  config = async (): Promise<ConfigResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.config());
  };
  pendingClaims = async (params: CamelCasedProperties<Extract<QueryMsg, {
    pending_claims: unknown;
  }>[\\"pending_claims\\"]>): Promise<PendingClaimsResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.pendingClaims(params));
  };
  claims = async (params: CamelCasedProperties<Extract<QueryMsg, {
    claims: unknown;
  }>[\\"claims\\"]>): Promise<ClaimsResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.claims(params));
  };
  allClaims = async (params: CamelCasedProperties<Extract<QueryMsg, {
    all_claims: unknown;
  }>[\\"all_claims\\"]>): Promise<AllClaimsResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.allClaims(params));
  };
  latestUnbonding = async (): Promise<LatestUnbondingResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.latestUnbonding());
  };
  totalLpPosition = async (): Promise<TotalLpPositionResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.totalLpPosition());
  };
  balance = async (params: CamelCasedProperties<Extract<QueryMsg, {
    balance: unknown;
  }>[\\"balance\\"]>): Promise<BalanceResponse> => {
    return this._query(AutocompounderQueryMsgBuilder.balance(params));
  };
  getAddress = async (): Promise<string> => {
    if (!this._moduleAddress) {
      this._moduleAddress = await this.accountQueryClient.getModuleAddress(this.moduleId);
    }

    return this._moduleAddress!;
  };
  connectSigningClient = (signingClient: SigningCosmWasmClient, address: string): AutocompounderAppClient => {
    return new AutocompounderAppClient({
      accountId: this.accountQueryClient.accountId,
      managerAddress: this.accountQueryClient.managerAddress,
      proxyAddress: this.accountQueryClient.proxyAddress,
      moduleId: this.moduleId,
      abstractClient: this.accountQueryClient.abstract.connectSigningClient(signingClient, address)
    });
  };
  _query = async (queryMsg: QueryMsg): Promise<any> => {
    return this.accountQueryClient.queryModule({
      moduleId: this.moduleId,
      moduleType: \\"app\\",
      queryMsg
    });
  };
}"
`;

exports[`IAutocompounderAppClient 1`] = `
"export interface IAutocompounderAppClient extends IAutocompounderAppQueryClient {
  accountClient: AbstractAccountClient;
  updateFeeConfig: (params: CamelCasedProperties<Extract<ExecuteMsg, {
    update_fee_config: unknown;
  }>[\\"update_fee_config\\"]>, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  deposit: (params: CamelCasedProperties<Extract<ExecuteMsg, {
    deposit: unknown;
  }>[\\"deposit\\"]>, fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  withdraw: (fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  compound: (fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
  batchUnbond: (fee?: number | StdFee | \\"auto\\", memo?: string, _funds?: Coin[]) => Promise<ExecuteResult>;
}"
`;

exports[`IAutocompounderAppQueryClient 1`] = `
"export interface IAutocompounderAppQueryClient {
  moduleId: string;
  accountQueryClient: AbstractAccountQueryClient;
  _moduleAddress: string | undefined;
  config: () => Promise<ConfigResponse>;
  pendingClaims: (params: CamelCasedProperties<Extract<QueryMsg, {
    pending_claims: unknown;
  }>[\\"pending_claims\\"]>) => Promise<PendingClaimsResponse>;
  claims: (params: CamelCasedProperties<Extract<QueryMsg, {
    claims: unknown;
  }>[\\"claims\\"]>) => Promise<ClaimsResponse>;
  allClaims: (params: CamelCasedProperties<Extract<QueryMsg, {
    all_claims: unknown;
  }>[\\"all_claims\\"]>) => Promise<AllClaimsResponse>;
  latestUnbonding: () => Promise<LatestUnbondingResponse>;
  totalLpPosition: () => Promise<TotalLpPositionResponse>;
  balance: (params: CamelCasedProperties<Extract<QueryMsg, {
    balance: unknown;
  }>[\\"balance\\"]>) => Promise<BalanceResponse>;
  connectSigningClient: (signingClient: SigningCosmWasmClient, address: string) => AutocompounderAppClient;
  getAddress: () => Promise<string>;
}"
`;
